### PR TITLE
[FLINK-4574] [kinesis] Strengthen fetch interval implementation in Kinesis consumer

### PR DESCRIPTION
**Pull request by tony810430**
I used Timer to implement it.

If "flink.shard.getrecords.intervalmillis" is set by default value, which is 0, the timer will schedule ShardConsumerFetcher once and run it forever.
If "flink.shard.getrecords.intervalmillis" is greater than 0, the timer will schedule ShardConsumerFetcher at a fixed ratio by using timer.scheduleAtFixedRate, which makes sure two consecutive function call would be a fixed interval.
But if the getRecords took too much time and couldn't be finished on time, ShardConsumerFetcher would log the warning and drop the next delayed task.

Ideally :
|----p1----|----p2----|----p3----|
|=====>...|====>......|====>......|
  task1.........task2.........task3

task2 is delayed by task1: task2 will be dropped
|----p1----|----p2----|----p3----|
|============>..... |====>......|
  task1...........................task3

### CODE REVIEW

**Comment by StephanEwen** [comment]
@tzulitai I think it would be great to get your input on this pull request, when you find the time.

**Comment by tzulitai** [comment]
I'm definitely planning to look at this over the next few days :) Currently quite overwhelmed right now.

Thanks for all your recent work on the Kinesis connector @tony810430, and very sorry for the late reviews. Please bear with me for a little while, I'll get back to the PRs soon ;)

**Comment by tzulitai** [comment]
Thanks for working on this @tony810430!

I'd like to discuss a bit on the design before a more detailed review for this PR. If I'm correct, with this PR we will be creating a separate shard consuming thread with the Timer even if the fetch interval is 0 (which is our default). In this case, especially that it's our default, this extra thread usage seems sub-optimal to me, and I'm wondering if there's a better design for this. What do you think?

**Comment by tony810430** [comment]
I think it would not be sub-optimal, because it still need to separate consuming thread and ingesting thread, which emits records to next operator. The reason for this implementation is because `Timer` can't support 0 period, so I have to make it run forever.

**Comment by tzulitai** [comment]
@tony810430 ok, thanks for your response and patience on the review.
I'll proceed with reviewing this and testing it against Kinesis either today or tomorrow.

File: flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/ShardConsumer.java [hunk 1]
```diff
@@ -154,42 +166,115 @@ public void run() {
...
+			this.runForever = runForever;
+		}
+
+		@Override
+		public void run() {
+			try {
+				do {
+					if (nextShardItr != null) {
+						// ignore to log this warning if runForever is true, since fetchIntervalMillis is 0
+						if (!runForever && this.scheduledExecutionTime() < lastFinishTime) {
```
Comment (tzulitai) [inline flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/ShardConsumer.java]: I think `scheduledExecutionTime ` return the time of the _actual execution time_, so this should always be false.
See https://docs.oracle.com/javase/7/docs/api/java/util/TimerTask.html#scheduledExecutionTime().

File: flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/ShardConsumer.java [hunk 2]
```diff
@@ -154,42 +166,115 @@ public void run() {
...
+							// If expected scheduled execution time is earlier than lastFinishTime,
+							// it seems that the fetchIntervalMillis might be short to finish the previous task.
+							LOG.warn("The value given for ShardConsumer is too short to finish getRecords on time. Please increase the value set in config \"{}\"", ConsumerConfigConstants.SHARD_GETRECORDS_INTERVAL_MILLIS);
+						} else {
+							GetRecordsResult getRecordsResult = shardConsumerRef.getRecords(nextShardItr, maxNumberOfRecordsPerFetch);
```
Comment (tzulitai) [inline flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/ShardConsumer.java]: Can we declare `GetRecordsResult getRecordsResult` outside of the do-while scope and reuse the variable?

File: flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/ShardConsumer.java [hunk 3]
```diff
@@ -154,42 +166,115 @@ public void run() {
...
+
+							if (getRecordsResult != null) {
+								// each of the Kinesis records may be aggregated, so we must deaggregate them before proceeding
+								List<UserRecord> fetchedRecords = deaggregateRecords(
```
Comment (tzulitai) [inline flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/ShardConsumer.java]: Same here: make `fetchedRecords` reusable?

File: flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/ShardConsumer.java [hunk 4]
```diff
@@ -203,6 +288,12 @@ private boolean isRunning() {
 		return !Thread.interrupted();
 	}
 
+	/** Called by created TimerTask to pass on errors. Only the first thrown error is set.
```
Comment (tzulitai) [inline flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/ShardConsumer.java]: I would simply reference `{@link ShardConsumerFetcher}` to be more specific.

File: flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/ShardConsumer.java (hunk 1 above)
Comment (tony810430) [inline flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/ShardConsumer.java]: The implementation for `scheduledExecutionTime` is 
```
public long scheduledExecutionTime() {
    synchronized(lock) {
        return (period < 0 ? nextExecutionTime + period : nextExecutionTime - period);
    }
}
```
and `nextExecutionTime` is updated by `nextExecutionTime + period`.

In `scheduleAtFixedRate` mode, `scheduledExecutionTime` will actually return the expected execution time, so I would use it to determine if the task was delayed.

**Comment by tony810430** [comment]
Hi, @tzulitai 

I think there is no better way if trying to use `Timer` to deal with `0` fetch interval. Even though I implement my own `Timer` scheduler, it just move the `runForever` inside from `ShardConsumerTask`. Because fetch interval is `0`, there is no way to get the next expected execution time and the solution would be running the function forever.

I think a better solution is defined non-positive fetch interval differs from positive fetch interval and would be executed in another way in the document. The implementation will be two parts: the original way remains the same without `runForever` in `ShardConsumerFetcherTask` and the other is for non-positive fetch interval implemented by using a Thread running `ShardConsumerFetcherTask::run()` forever.

It is more reasonable for me to distinguish these two configuration and implement them by using two ways in `ShardConsumer::run()`. What do you think?

**Pull request closed without merging**